File: nezal-ar/public/base-defender/js/game/isometric/moving_object.js
var MovingObject = Class.create({
	 coords : null,
	 moving : false ,
	 rotating : false,
	 targetAngle :0,
	 angle : 0,
	 speed : 0.7,
	 distanceToNextTile : 0,
	 noOfStates : 8,
	 state : 0,
	 counter : 0,
	 initialize : function(game,x,y){
	 	this.game = game
		this.coords = {}
	 	this.coords.x = x
		this.coords.y = y
		this.shadow = {xdim: 40, ydim: 40, zdim: 0, imgWidth: 50, imgHeight: 31, angle:0};
		this.shadow.coords = this.coords
		this.movingPath = []
		this.game.scene.push(this);
	 },
	 
	 tick : function(){
	 		this.counter++
	 		if(this.movingPath.length>0){
			if(this.targetAngle!=this.angle){
				this.changeAngle();
				return
			}
			var values = Map.value(this.movingPath[this.movingPath.length-1].x,this.movingPath[this.movingPath.length-1].y)
			if(!this.moving){
				this.moving = true
				this.distanceToNextTile = Util.distance(this.coords.x,this.coords.y,values[0],values[1])  
				this.targetAngle = Map.getDirection(this.coords.x,this.coords.y,values[0],values[1])
			}
			//this.state = this.calculateCurrentState(values)
			var movements = Util.getNextMove(this.coords.x,this.coords.y,values[0],values[1],this.distanceToNextTile/24)
			if(this.counter%3==0)this.state = (this.state+1)%8
			this.coords.x+=movements[0]
			this.coords.y+=movements[1]
			var mapValues = Map.tileValue(this.coords.x,this.coords.y)
			if(this.coords.x == values[0] && this.coords.y == values[1]){
				this.movingPath.pop()
				if(this.movingPath.length>0){
					values = Map.value(this.movingPath[this.movingPath.length-1].x,this.movingPath[this.movingPath.length-1].y)
					this.targetAngle = Map.getDirection(this.coords.x,this.coords.y,values[0],values[1])
					//this.angle = Map.getDirection(this.x,this.y,values[0],values[1])
				}
			}
			if(this.movingPath.length == 0){
				this.moving = false
				this.state = 0
			} 
		}
		else {
			var rand = Math.random()
			if (rand <= 0.05) {
		  	var x = Math.round(game.scene.map.x + game.scene.map.viewWidth * Math.random())
		  	var y = Math.round(game.scene.map.y + game.scene.map.viewHeight * Math.random())
		  	Map.moveObject(this, x, y)
	  	}
		}
	},
	changeAngle : function(){
		if(this.angle > this.targetAngle){
			if((this.targetAngle+8) - this.angle < this.angle - this.targetAngle){
				this.angle = (this.angle+1) % 8 
			}else{
				this.angle = (this.angle+7) % 8
			}
		}else{
			if(this.targetAngle - this.angle < (this.angle+8) - this.targetAngle){
				this.angle = (this.angle+1) % 8 
			}else{
				this.angle = (this.angle+7) % 8
			}
		}			
	},
	calculateCurrentState : function(values){
		var numirator = (this.distanceToNextTile - Util.distance(this.coords.x,this.coords.y,values[0],values[1]))*this.noOfStates
		return Math.round(numirator/this.distanceToNextTile)
	}
})

File: nezal-ar/app/models/games/base_defender/isometric/moving_object.rb
module BD
  class MovingObject
    def initialize(map,x,y)
      @coords = nil
      @moving = false 
      @rotating = false
      @target_angle =0
      @angle = 0
      @speed = 0.7
      @random_move = false
      @distance_to_next_tile = 0
      @no_of_states = 8
      @state = 0
      @tick_counter = 0
      @coords = {}
      @coords['x'] = x
      @coords['y'] = y
      @moving_path = []
      @map = map
      @target_point = nil
    end
    attr_accessor :coords, :moving, :rotating, :target_angle, :angle, :speed, :random_move, :distance_to_next_tile, 
    :no_of_states, :state, :tick_counter, :map, :moving_path, :target_point
    def tick
      @tick_counter= @tick_counter+1 
      if(@target_angle!=@angle)
          change_angle()
          return
        end
      if(!@moving_path.nil?)
        if(@moving_path.length>0)
          values = @map.value(@moving_path[@moving_path.length-1].x,@moving_path[@moving_path.length-1].y)
          if(!@moving)
            @moving = true
            @distance_to_next_tile = Util.distance(@coords['x'],@coords['y'],values[0],values[1])  
            @target_angle = @map.get_direction(@coords['x'],@coords['y'],values[0],values[1])
          end
          movements = Util.get_next_move(@coords['x'], @coords['y'], values[0], values[1], @speed)
          @state = (@state+1)%8 if(@tick_counter%3==0)
          @coords['x']+=movements[0]
          @coords['y']+=movements[1]
          map_values = @map.tile_value(@coords['x'], @coords['y'])
          if(@coords['x'] == values[0] && @coords['y'] == values[1])
            @moving_path.pop()
            if(@moving_path.length>0)
              values = @map.value(@moving_path[@moving_path.length-1].x, @moving_path[@moving_path.length-1].y)
              @target_angle = @map.get_direction(@coords['x'], @coords['y'], values[0], values[1])
            end
          end
         elsif(@moving_path.length == 0)
          if(@moving)
            if ((@coords['x'] - @target_point['x']).abs <0.01 && (@coords['y'] - @target_point['y']).abs <0.01) 
            @moving = false
            @state = 0
          else
            @target_angle = @map.get_direction(@coords['x'],@coords['y'],@target_point['x'], @target_point['y'])
            movements = Util.get_next_move(@coords['x'], @coords['y'], @target_point['x'], @target_point['y'], @speed)
            @coords['x']+=movements[0]
            @coords['y']+=movements[1]
          end
         end
        end
      end
    end
  
    def change_angle 
      if(@angle > @target_angle)
        if((@target_angle+8) - @angle < @angle - @target_angle)
          @angle = (@angle+1) % 8 
        else
          @angle = (@angle+7) % 8
        end
      else
        if(@target_angle - @angle < (@angle+8) - @target_angle)
          @angle = (@angle+1) % 8 
        else
          @angle = (@angle+7) % 8
        end
      end     
    end
  
    def calculate_current_state values
      numirator = (@distance_to_next_tile - Util.distance(@coords['x'],@coords['y'],values[0],values[1]))*@no_of_states
      return (numirator/@distance_to_next_tile).round
    end
  end
end
